### ISSUE TITLE
How to write typescript (or at least es6) in the gatsby-config.js? I need to import a ts file but fails.

### ISSUE THREAD

**Issue by hackhat** [gatsbyjs/gatsby#10452]

**Comment by pieh** [comment]
`gatsby-config.js` as well as `gatsby-node.js` and `gatsby-ssr.js` are running on Node.js "as is" (without any additional processing), so they don't support typescript.

If you would like to use typescript in those files you kind of can do that, but you will need to compile it to `.js` in the end - you probably can automate it in your project, but Gatsby on itself is not doing that.

**Comment by hackhat** [comment]
yeah, I got it. How can I run gatsby like a normal node.js script? I really dislike that I need to use `gatsby` to run my app. I'd like to run it with `ts-node-dev`

**Comment by sidharthachatterjee** [comment]
You should be able to run `ts-node-dev ./node_modules/.bin/gatsby develop`

**Comment by hackhat** [comment]
Already tried but it says that can't find the worker or something similar. Looks like I can't pass the `develop` argument to gatsby properly or it's looking at the wrong place.

Anyway, do you know why has been decide to use `gatsby` instead of node? what are the benefits?

**Comment by sidharthachatterjee** [comment]
The `gatsby` cli is meant for ease of use and to make bootstrapping, developing and building a new Gatsby app simple

The `cli` is a pretty thin wrapper though and you may skip it altogether and write your own custom wrapper if you like 

The entry point for `build` is at https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/commands/build.js

**Comment by jonniebigodes** [comment]
@sidharthachatterjee issuing the command you mentioned is not the way to go, you basically have a "server" watching for files and then calling another on top of that, as gatsby is using under the hood another. And this can cause all sorts of headaches, more even that, for the actual changes to be picked up by the `gatsby-config.js` file you would have to restart the process every time, bluntly saying hot reload is for actual content not setup. @hackhat i'm sorry but this is one of those situations in which you have to make the development a little more cumbersome. Typescript can be used with gatsby, but not with the actual config file you wanted. Without further information on what the actual import you want to add to the configuration file and it's purpose and picking up on what pieh said, one approach that you could do is transform the ts file to a folder with the .js extension and then require the js file into `gatsby-config.js` file and then issue `gatsby develop` for instance, and see if everything is working properly.

**Comment by sidharthachatterjee** [comment]
@jonniebigodes Yup, I meant `ts-node ./node_modules/.bin/gatsby develop` as opposed to `ts-node-dev`

**Comment by maximgeerinck** [comment]
Having the same issue. I have a file createPages.ts which i have to require from gatsby-node.js (cause createPages goes over components written in tsx).

**Comment by jedrichards** [comment]
Another option is to use `@babel/register` at the top of the file you want to import TS into:

```
require('@babel/register')({
  extensions: ['.js', '.ts', '.tsx', '.jsx'],
  presets: [
    require.resolve('babel-preset-gatsby'),
    [
      require.resolve('@babel/preset-typescript'),
      { isTSX: true, allExtensions: true },
    ],
  ],
})
```

**Comment by assainov** [comment]
Here are two ways you can _write_ typescript in gatsby-config:
1. Create gatsby-config.ts and use tsc to target the file and emit gatsby-config.js in the root of the project
2. Use ts-node to allow importing typescript modules into gatsby-config.js.

For a more detailed explanation, check out this blogpost: [Converting Gatsby Config and Node API to TypeScript](https://www.extensive.one/converting-gatsby-config-and-node-api-to-typescript/)

**Comment by LekoArts** [comment]
Please go to https://github.com/gatsbyjs/gatsby/discussions/34613 and you can use TypeScript with Gatsby. By the time you're reading this it's also possible that the feature already was shipped in Gatsby 4. If that is the case the RFC will be marked as done.

**Comment by LekoArts** [comment]
Future googlers: This is available in 4.9. https://www.gatsbyjs.com/docs/reference/release-notes/v4.9/#support-for-typescript-in-gatsby-config-and-gatsby-node
